A47-xds-federation.md

```diff
@@ -0,0 +1,693 @@
+new-style names, it will be the `id` part of the `xdstp:` URI (i.e., the
+URI path with the resource type prefix stripped off) and the context
+parameters.
+
+For example, a resource with old-style name `server.example.com` and
+type Listener might be stored like this:
+
+```
+{authority "old:" =>
+  {resource_type Listener =>
+    {"server.example.com" => <resource>}
+  }
+}
+```
+
+But a resource with new-style name
+`xdstp://xds.authority.com/envoy.config.listener.v3.Listener/server.example.com?foo=bar`
+might be stored like this:
+
+```
+{authority "xdstp:xds.authority.com" =>
+  {resource_type Listener =>
+    {("server.example.com", {"foo"=>"bar"}) => <resource>}
+  }
+}
+```
+
+New-style resource names used as cache keys must be
+[normalized](https://github.com/cncf/xds/blob/main/proposals/TP1-xds-transport-next.md#normalization).
+In particular, this means sorting the context params in normal lexicographic
+order.
+
+If a new-style name has duplicate values for the same context param, the
+behavior is undefined.  The easiest thing is probably to just use the
+last value seen, since this allows the implementation to just populate
+the map by iterating over the query params.
+
+##### Watcher Changes
+
+When a caller starts a watch on an old-style resource name (one that
+does not start with `xdstp:`), the `XdsClient` will behave essentially as
+it does today, using the empty string as the authority name in the
+resource cache.
+
+When a caller starts a watch on a new-style resource name (one that
+starts with `xdstp:`), the `XdsClient` will parse the URI.  It will use
+the authority of the URI as the primary key in the resource cache (using
+the empty string if the URI has no authority).  It will canonify any
+context parameters that are part of the resource name (i.e., removing
+duplicate keys and sorting keys in lexicographic order).  It will then
+check the cache to see if the resource is already present, in which case
+it will be returned immediately.  Otherwise, it will send a message
+subscribing to that resource, using the appropriate xDS server from the
+authority.
+
+#### LRS Server Representation
+
+We currently represent the LRS server name in several places:
+- In the `CdsUpdate` struct returned by the `XdsClient` CDS watcher.
+- In the LB policy configs for the `xds_cluster_resolver` and
+  `xds_cluster_impl` LB policies.
+- In the `XdsClient` load reporting APIs.
+
+Currently, the name is represented as just a string, and it basically
+has only two values: unset, meaning that load reporting is not enabled,
+or the empty string, meaning to use the same server as we're using for
+ADS.  The latter meaning worked fine when there was only one possible
+server for us to talk to, but in the presence of federation, there can
+be multiple servers, and we need a way to indicate which one we mean.
+
+In the general case, we actually need more than just the server name for
```

Comment:
Will we report loads to a server that's not already an `authority` in the bootstrap config?
(We only support `self` today, and there's no xDS `ConfigSource` to specify a complete authority-config or `XdsServer`).

I'm thinking if we can just specify the `authority` name here, and the xdsclient can look up for the real config in the bootstrap.

Reply:
The `XdsServer` message here is basically our equivalent of the xDS `ConfigSource` proto.  If we did in the future decide to accept arbitrary `ConfigSource`s, it could specify a server other than one already listed in the bootstrap file, in which case we'd populate the `XdsServer` message from the `ConfigSource` (the [`target_uri` field](https://github.com/envoyproxy/envoy/blob/9d25410796882dabc4952ca26cab608930693e84/api/envoy/config/core/v3/grpc_service.proto#L235) would map to our `server_uri` field, the [`channel_credentials` field](https://github.com/envoyproxy/envoy/blob/9d25410796882dabc4952ca26cab608930693e84/api/envoy/config/core/v3/grpc_service.proto#L237) would map to our `channel_creds` field, and we'd seed our `server_features` field from the [`transport_api_version` field](https://github.com/envoyproxy/envoy/blob/9d25410796882dabc4952ca26cab608930693e84/api/envoy/config/core/v3/config_source.proto#L82)).  Even though we're not going to allow that today, we are trying to set the stage for doing so later.

I don't think we want to use the authority here, for several reasons:
- The authority as a concept is really an xDS-specific thing, and it makes sense only in the context of new-style xDS resource names.  LRS is not the same as xDS (it's a companion protocol to xDS but not identical to it), and it does not use resource names to specify the LRS server itself.
- If we did use an xDS authority name to specify the LRS server, then that would cause problems later if we wanted to support an LRS server that is different from the xDS server.
- In the short term, there will be many configs for which the LRS server is also the xDS server but the only xDS server configured in the bootstrap file is the global one, not one using an authority.  If we used an authority to specify the name, then we would have no way to indicate to use the global xDS server from the boostrap file.

Reply:
> > The authority as a concept is really an xDS-specific thing, and it makes sense only in the context of new-style xDS resource names. LRS is not the same as xDS (it's a companion protocol to xDS but not identical to it), and it does not use resource names to specify the LRS server itself.
> 
> Right now LRS is only "return to sender", so we aren't specifying the LRS server as much as the CDS source.

Even if we were phrasing it as "the CDS source", we still wouldn't want to specify it as an authority, because authority as a concept is completely irrelevant to LRS.

To say this another way, I think that in the fullness of time, we will want to move the LRS functionality out of the `XdsClient` and into a separate interface.  At that point, it will not make sense to specify the LRS server as an authority, because the authority will be a concept relevant to the `XdsClient`, not to the LRS interface.

> 
> > If we did use an xDS authority name to specify the LRS server, then that would cause problems later if we wanted to support an LRS server that is different from the xDS server.
> 
> What problems does it cause? We'd just have to add more code to support it and get rid of our current simple-string approach. But doing that in the future is just as easy as doing it now, except that we don't have to do it now.

We need to change the representation now to support federation anyway.  I don't think it makes sense to change to something that I'm confident will not work in the long run.  It seems better to change to something that I think will continue to work in the future, so that we minimize the chances of having to go change it again later.

> 
> In truth, it isn't clear this is the best future approach either. To avoid security problems, we have the restriction in the "XdsClient Load Reporting APIs" section that the config used here is already known from the bootstrap file. 

I think that the current security restrictions about requiring the LRS server to be one of the xDS servers in the bootstrap file is probably a short-term restriction.  I think that in the long term, we are going to have to find a way to solve the problem of being able to trust the control plane, because if we don't do that, we're going to start facing increasingly serious security problems that we can't work around easily.  These are problems that are larger in scope than simply how to specify the name of another control plane to talk to.


> It seems far easier if this is just use a "cookie", and that cookie could be the authority for SelfConfigSource. In the future we can generate an id (random, sequential, or hash) for xds-provided ConfigSources and store them in a map. That is obviously secure, trivial to plumb, and easier to implement in load reporting (vs a linear scan of all servers in bootstrap with a very specific equals check).

I think that approach seems like an option only because we happen to have xDS and LRS both handled via the `XdsClient` interface today.  I don't think it will make sense in the long run.

> 
> > In the short term, there will be many configs for which the LRS server is also the xDS server but the only xDS server configured in the bootstrap file is the global one, not one using an authority. If we used an authority to specify the name, then we would have no way to indicate to use the global xDS server from the boostrap file.
> 
> There's lots of disallowed charactes to use `&%/oldstyle`. Or if that bothers you too much, just use a different field name for old vs new. This seems a very small problem to instead argue for doing this much more invasive change within this design.

I think this would actually be more special-casey and more complicated than the current design.  It bakes in the assumption that you have to know whether it's an old or new name, which in principle should have no relevance to LRS, because that's a purely xDS-relevant concept.

I think that expressing the LRS server name directly is actually quite clean and general-purpose.  The only special case we have is that the `XdsClient` currently requires that the name actually be one of the ones in the bootstrap file, and that special case is well-contained and will be easy to remove in the future.

Reply:
> Even if we were phrasing it as "the CDS source", we still wouldn't want to specify it as an authority, because authority as a concept is completely irrelevant to LRS.

Well, to the perspective of the receiver in the service config, it'd just be a "cookie" or "server id". It'd be treated as opaque.

> We need to change the representation now to support federation anyway.

It is a string today. I was suggesting leaving it as a string. The only difference is lrs would need to pass that string into xds client when reporting.

> To say this another way, I think that in the fullness of time, we will want to move the LRS functionality out of the XdsClient and into a separate interface.

When we do that, there's a serious security issue to resolve. We _can't_ treat this as a general-purpose `ConfigSource`. Somehow xds client and the lrs client would need to coordinate (maybe through a third class, LrsApprovedServerContainer) to make sure the server config is trusted. Since they have to directly coordinate, we can pass an id or "handle" instead of the full configuration.

I'd also accept using "server1", "server2", etc, and not using the authority. Any string generation works; I can easily live without using the authority string, although I do think it'd be the easiest for implementation and debugging. We can change what the string is at any time and only the component that generated the string should care (i.e., xds client).

> I think that the current security restrictions about requiring the LRS server to be one of the xDS servers in the bootstrap file is probably a short-term restriction. I think that in the long term, we are going to have to find a way to solve the problem of being able to trust the control plane,

That's not at all what I was talking about. The design here and my proposal both work fine if we trust the control plane. The point is that "ConfigSources must be trusted, and service config isn't trusted", so we can give such sources a name when we get them (from bootstrap or control plane) and use that name. We don't need to provide the full config in service config because LRS _isn't allowed_ to use any arbitrary config; at any given time there's a fixed number of ConfigSources it can use.

> I think that approach seems like an option only because we happen to have xDS and LRS both handled via the XdsClient interface today

In all futures, xds client and lrs client have to coordinate for security. Unless we wanted to start signing config snippets in service config, which still requires coordination for the key.

>> There's lots of disallowed charactes to use `&%/oldstyle`
> I think this would actually be more special-casey and more complicated than the current design. It bakes in the assumption that you have to know whether it's an old or new name

Again, only xds client has to care (or a new class that mediates between xds client and lrs client). I'm not wild about a separate field, but it certainly is "less complex" than XdsServer solution. The special string seems very minor, especially since such a name may already exist (`old:` mentioned elsewhere).

> The only special case we have is that the XdsClient currently requires that the name actually be one of the ones in the bootstrap file, and that special case is well-contained and will be easy to remove in the future.

That special case can never be removed. It can be enhanced to be "in the bootstrap file or from control plane", but we must always have the validation check in place.

> in principle should have no relevance to LRS, because that's a purely xDS-relevant concept.

Well, right now security is only designed for xDS and ConfigSource must be trusted. But if we consider a gRPC-LB and RLS approach, LRS could become safe if CallCredentials were stripped... Mostly? Arbitrary ChannelCredentials configuration makes me nervous with its power. Ummm... is it safe to send load reports to an attacker's server, even if the credentials don't get leaked? The Node proto, cluster names, localities, or ORCA metrics may be poor to send to an attacker?

At the very least, LRS would need special xDS vs non-xDS logic. I'm sympathetic to this idea, but it seems premature to claim that XdsServer is well suited to the task.

Reply:
Just to be clear after the book: I can live with the XdsServer message in this design. It seems unnecessary, but it's not _that_ bad and all this is internal. If it is a misstep it wouldn't be hard to undo.

Reply:
Eric, you raise a good point about the issue of trusting the service config, as opposed to trusting the xDS control plane.  I think we have some disagreements about whether or not the service config should be considered untrusted in the general case, and it will require deeper discussion to come to consensus on that in the long term.  But as you say, I don't think this is a blocker for the current design.

Let's continue to discuss this as we move forward.